prov/mlx/src/mlx_fabric.c

```diff
@@ -88,5 +171,37 @@ int mlx_fabric_open(
 	fabric_priv->u_fabric.fabric_fid.ops = &mlx_fabric_ops;
 	*fabric = &(fabric_priv->u_fabric.fabric_fid);
 
+	if (mlx_descriptor.use_ns) {
+		if (!mlx_descriptor.localhost)
+			mlx_descriptor.localhost = mlx_local_host_resolve();
+
+		if (!mlx_descriptor.localhost) {
+			FI_INFO(&mlx_prov, FI_LOG_CORE,
+				"Unable to resolve local host address:\n"
+				"\t - unable to start NS\n"
+				"\t - Please try MLX-adress format");
+			return -FI_EINVAL;
+		}
+
+		struct util_ns_attr ns_attr = {
+			.ns_hostname = mlx_descriptor.localhost,
+			.ns_port = (int)mlx_descriptor.ns_port,
+			.name_len = FI_MLX_MAX_NAME_LEN,
+			.service_len = sizeof(short),
+			.service_cmp = mlx_ns_service_cmp,
+			.is_service_wildcard = mlx_ns_is_service_wildcard,
+		};
+
+		status = ofi_ns_init(&ns_attr,
+				  &mlx_descriptor.name_serv);
+		if (status) {
+			FI_INFO(&mlx_prov, FI_LOG_CORE,
+			        "ofi_ns_init returns %d\n", status);
+			free(fabric_priv);
+			return status;
+		}
+		ofi_ns_start_server(&mlx_descriptor.name_serv);
```

Comment:
This is a large segment of code that would be better isolated into a new function.